File: collect_app/src/main/java/org/odk/collect/android/widgets/range/RangeIntegerWidget.java
/*
 * Copyright 2017 Nafundi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.odk.collect.android.widgets.range;

import android.annotation.SuppressLint;
import android.content.Context;

import org.odk.collect.android.formentry.questions.QuestionDetails;

@SuppressLint("ViewConstructor")
public class RangeIntegerWidget extends RangeBaseWidget {
    public RangeIntegerWidget(Context context,
                              QuestionDetails prompt,
                              Dependencies dependencies) {
        super(context, prompt, dependencies, true);
    }
}

File: collect_app/src/main/java/org/odk/collect/android/widgets/range/RangeIntegerWidget.kt
package org.odk.collect.android.widgets.range

import android.annotation.SuppressLint
import android.content.Context
import android.view.View
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView
import org.javarosa.core.model.data.IAnswerData
import org.javarosa.core.model.data.IntegerData
import org.javarosa.form.api.FormEntryPrompt
import org.odk.collect.android.formentry.questions.QuestionDetails
import org.odk.collect.android.widgets.QuestionWidget
import org.odk.collect.androidshared.ui.ComposeThemeProvider.Companion.setContextThemedContent
import org.odk.collect.androidshared.ui.ToastUtils

@SuppressLint("ViewConstructor")
class RangeIntegerWidget(
    context: Context,
    prompt: QuestionDetails,
    dependencies: Dependencies
) : QuestionWidget(
        context,
        dependencies,
        prompt
    ) {
    private var rangeSliderState by mutableStateOf(RangeSliderState.fromPrompt(formEntryPrompt))
    private var shouldSuppressFlingGesture = false

    init {
        render()
    }

    override fun onCreateWidgetView(context: Context, prompt: FormEntryPrompt, answerFontSize: Int): View {
        return ComposeView(context).apply {
            setContextThemedContent {
                RangeSlider(
                    rangeSliderState,
                    onValueChanging = { shouldSuppressFlingGesture = it },
                    onValueChangeFinished = {
                        rangeSliderState = it
                        widgetValueChanged()
                    },
                    onRangeInvalid = {
                        ToastUtils.showLongToast(org.odk.collect.strings.R.string.invalid_range_widget)
                    }
                )
            }
        }
    }

    override fun getAnswer(): IAnswerData? {
        return rangeSliderState.realValue?.let {
            IntegerData(it.toInt())
        }
    }

    override fun setOnLongClickListener(l: OnLongClickListener?) = Unit

    override fun shouldSuppressFlingGesture() = shouldSuppressFlingGesture

    override fun clearAnswer() {
        rangeSliderState = rangeSliderState.copy(sliderValue = null)
        widgetValueChanged()
    }
}
